File: Java/Classes/Polynom.java
package Classes;

import java.util.regex.Pattern;
import java.awt.geom.Point2D;
import java.util.Vector;

public class Polynom extends BaseClass
{
    private Vector<Point2D> list;
    
    private void sort()
    {
	
        for (int j = 1; j < list.size(); ++j)
	    {
		Point2D key = list.get(j);
		int i = j - 1;
		while (i >= 0 && list.get(i).getY() <= key.getY())
		    {
			list.set(i + 1, list.get(i));
			i--;
		    }
		list.set(i + 1, key);
	    }
    }
    
    private static Point2D converti(String a)
    {
        int i = 0;
	StringBuilder ax = new StringBuilder(a);
        if (ax.length() == 1 && ax.charAt(0) == 'x') //monomio = x
            return new Point2D.Double(1, 1);
        else
	    {
		if (ax.charAt(0) == 'x') //monomio = x^b
		    {
			a = a.replace("x", "");
			a = a.replace("^", "");
			return new Point2D.Double(1, Double.valueOf(a));
		    }
		else
		    {
			StringBuilder aux = new StringBuilder(); //aux è il coefficiente
			while (i < ax.length() && ax.charAt(i) != 'x')
			    {
				aux.append(ax.charAt(i));
				i++;
			    }

			if ((aux.charAt(0) == '-' || aux.charAt(0) == '+') && aux.length() == 1) // monomio = -x^b o +x^b
			    aux.append('1');
			
			ax.delete(0,i);
			
			if (ax.length() == 0) //monomio = a
			    return new Point2D.Double(Double.valueOf(aux.toString()), 0);
			
			a = ax.toString();
			a = a.replace("x", "");
			a = a.replace("^", "");
			if (a.length() == 0) //monommio = ax
			    return new Point2D.Double(Double.valueOf(aux.toString()), 1);
			
			return new Point2D.Double(Double.valueOf(aux.toString()), Integer.valueOf(a)); //monomio = ax^b
		    }
	    }
    }
    
    bool Polynom::polyVerify(String input)
    {
	regex number("((\\+|-)?[[:digit:]]*(x|(x[^][[:digit:]]+))?)+");
	return regex_match(input.toStdString(),number);    
    }
    
    public Polynom(String poly, String name)
    {
        super(name);
        list = new Vector<>();
        int i = 0;
        StringBuilder p = new StringBuilder(poly);
        while (i < p.length())
	    {
		StringBuilder aux = new StringBuilder();
		
		if (p.charAt(i) == '+' || p.charAt(i) == '-')
		    {
			aux.append(p.charAt(i));
			i++;
		    }
		
		while (i < p.length() && p.charAt(i) != '+' && p.charAt(i) != '-')
		    {
			aux.append(p.charAt(i));
			i++;
		    }
		
		Point2D point = converti(aux.toString());
		list.addElement(point);
	    }
        sort();
    }
    
    public Polynom(Polynom p)
    {
        super(p.getName());
        list = p.list;
    }
    
    public Polynom(Vector<Point2D> p, String name)
    {
        super(name);
        list = p;
    }

    public Polynom()
    {
        super("null");
        list = new Vector<>();
    }

    public Polynom sum(Polynom p)
    {
        if (list.isEmpty())
            return p;
	
        Vector<Point2D> aux = new Vector<>();
        int j = 0, k = 0;
	
        while (j < list.size() && k < p.list.size())
        {
            if (list.get(j).getY() == p.list.get(k).getY())
		{
		    if (list.get(j).getX() + p.list.get(k).getX() != 0.)
			aux.addElement(new Point2D.Double(list.get(j).getX() + p.list.get(k).getX(), list.get(j).getY()));
		    k++;
		    j++;
		}
            else
		{
		    if (list.get(j).getY() > p.list.get(k).getY())
			{
			    aux.addElement(list.get(j));
			    j++;
			}
		    else if (list.get(j).getY() < p.list.get(k).getY())
			{
			    aux.addElement(p.list.get(k));
			    k++;
			}
		}
        }

        while (j < list.size())
	    {
		aux.addElement(new Point2D.Double(list.get(j).getX(), list.get(j).getY()));
		j++;
	    }
	
        while (k < p.list.size())
	    {
		aux.addElement(new Point2D.Double(p.list.get(k).getX(), p.list.get(k).getY()));
		k++;
	    }
	
        Polynom a = new Polynom(aux, "Result");
	
        if (a.list.isEmpty())
            a.list.addElement(new Point2D.Double(0, 0));
	
        return a;
    }
    
    public Polynom minum(Polynom p)
    {
        if (list.isEmpty())
            return p;
	
        return  this.sum( p.mult(new Polynom("-1","")) ); // a - b = a + (-1)*b
    }
    
    public Polynom prod(Point2D term)
    {
        Vector<Point2D> aux=new Vector<>();
        for(int i=0; i<list.size(); i++)
	    {
		double coef = list.get(i).getX() * term.getX();
		double exp = list.get(i).getY() + term.getY();
		aux.addElement(new Point2D.Double(coef, exp));
	    }
        return new Polynom(aux,"");
    }
    
    public Polynom mult(Polynom p)
    {
        Polynom aux=new Polynom();
	
        for(int j=0; j<list.size(); j++)
	    {
		Polynom tmp = p.prod(this.list.get(j));
		aux=aux.sum(tmp);
	    }
        aux.setName("Result");
        return aux;
    }
    
    public Polynom div(Polynom p)
    {
	int i=0;
	while(i<list.size())
	    {
		if(i+1<list.size() && list.get(i).getY()-list.get(i+1).getY() != 1)
		    list.insert(i+1,new Point2D.Double(0,list.get(i).getY()-1));
		i++;
	    }
	
	Polynom q;
	while(p.list[0].second<=this->list[0].second)
	    {
		q.list.append(qMakePair(list[0].first/p.list[0].first,list[0].second-p.list[0].second));
		*this=*this-(p*q.list.last());
	    }
	q.setName("Result");
	this->setName("Resto");
	return q;
    }
    
    public String print()
    {
        StringBuilder aux=new StringBuilder();
        aux.append(super.print()+"= ");
        for(int i=0; i<list.size(); ++i)
	    {
		if(list.get(i).getX() > 0.0 && i != 0) //aggiunge + se coefficiente positivo
		    aux.append("+");
		if(list.get(i).getY() == 0) // esponente = 0 non aggungo la x
		    aux.append(String.valueOf(list.get(i).getX()));
		else //monomio del tipo ax^b
		    {
			if(list.get(i).getX() == -1. )
			    aux.append('-');
			else if(list.get(i).getX() != 1.0) //se coefficiente==1 non lo considero
			    aux.append(String.valueOf(list.get(i).getX()));
			
			aux.append("x");
			if(list.get(i).getY() != 1) // aggiungo l'esponente se != da 1
			    aux.append("^"+String.valueOf((int)list.get(i).getY()));
		    }
	    }
        return aux.toString();
    }
}

File: polynom.cpp
#include "polynom.h"
#include <cstdlib>
#include <typeinfo>
#include "exception.h"
#include <regex>
using std::regex;

#include <iostream>

QPair<double,int> Polynom::converti(QString a) //conversione da QString a qMakePair
{
  int i=0;
  if(a.size()==1 && a[0]=='x') //monomio = x
    return qMakePair(1,1);
  else
    {
      if(a[i]=='x') //monomio = x^b
        {
          a.remove('x');
          a.remove('^');
          return qMakePair(1,a.toInt());
        }
      else
        {
          QString aux;
          while(a[i]!='x' && i<a.size()) //aux è il coefficiente
            {
              aux.append(a[i]);
              ++i;
            }

          if((aux[0]=='-' || aux[0]=='+') && aux.size()==1) // monomio = -x^b
            aux.append('1');

          a.remove(0,i); //rimuovo i primi 'i' elementi cioè il coefficiente

          if(a.size()==0) //monomio = a
            return qMakePair(aux.toDouble(),0);

          a.remove('x');
          a.remove('^');

          if(a.size()==0) //monommio = ax
            return (qMakePair(aux.toDouble(),1));

          return qMakePair(aux.toDouble(), a.toInt()); //monomio = ax^b
        }
    }
}

void Polynom::sort()  //ordina in modo decrescente
{
  for (int j=1; j<list.size(); ++j)
    {

      QPair<double,int> key = list[j];
      int i=j-1;
      while(i>=0 && list[i].second<=key.second)
        {
          list[i+1]=list[i];
          i--;
        }

      list[i+1]=key;
    }
}


Polynom::Polynom(QString poly, QString name) : BaseClass (name)
{
  regex valid("((\\+|-)?[[:digit:]]*(x|(x[^][[:digit:]]+))?)+");
  if(!regex_match(poly.toStdString(),valid))
    {
      this->BaseClass::~BaseClass();
      throw InvalidPolynom();
    }

  int i=0;
  while(i<poly.size())
    {
      QString aux;
      if(poly[i]=='+' || poly[i]=='-')
        {
          aux.append(poly[i]);
          ++i;
        }

      while(i<poly.size() && poly[i]!='+' &&  poly[i]!='-')
        {
          aux.append(poly[i]);
          ++i;
        }

      QPair<double,int> p = converti(aux);
      list.push_back(p);
    }
  sort();
}

Polynom::Polynom(const Polynom& p) : BaseClass (p.getName()), list(p.list) {}

Polynom::Polynom(QVector<QPair<double, int> > p, QString name) : BaseClass (name), list(p) {}

Polynom::Polynom() : BaseClass(" "), list() { list .push_back(qMakePair(0,0)); }

Polynom* Polynom::operator+(const BaseClass& p) const
{
  Polynom a = static_cast<const Polynom&>(p);

  if(typeid(*this) != typeid(p))
    throw InvalidOperation();

  QVector<QPair<double,int>> aux;
  int j=0, k=0;
  while(j < list.size() && k < a.list.size())
    {
      if(list[j].second == a.list[k].second)
        {
          if(list[j].first + a.list[k].first != 0. )
            aux.push_back(qMakePair(list[j].first + a.list[k].first , list[j].second));
          k++;
          j++;
        }
      else
        {
          if(list[j].second > a.list[k].second)
            {
              aux.push_back(QPair<double,int>(list[j]));
              j++;
            }
          else
            if(list[j].second < a.list[k].second)
              {
                aux.push_back(QPair<double,int>(a.list[k]));
                k++;
              }
        }
    }

  //nel caso in cui il numero dei monomi dei due polinomi non sia lo stesso devo aggiungere quelli "mancanti" al risultato
  while(j < list.size())
    {
      aux.push_back(qMakePair(list[j].first,list[j].second));
      j++;
    }

  while(k < a.list.size())
    {
      aux.push_back(qMakePair(a.list[k].first,a.list[k].second));
      k++;
    }

  if(aux.empty())
    aux.push_back(qMakePair(0,0));

  return new Polynom(aux,"Result");
}

Polynom* Polynom::operator-(const BaseClass& p) const
{
  Polynom a = static_cast<const Polynom&>(p);

  if(typeid(*this) != typeid(p))
    throw InvalidOperation();

  return new Polynom(*(*(a * Polynom("-1","")) + *this ));

}

Polynom Polynom::molt(const QPair<double,int>& term) const
{
  QVector<QPair<double,int>> aux;
  for(int i=0; i<list.size(); ++i)
    {
      double coef=list[i].first * term.first;
      int exp=list[i].second + term.second;
      aux.push_back(qMakePair(coef, exp));
    }
  return Polynom(aux,"");
}

Polynom* Polynom::operator*(const BaseClass& p) const
{
  if(typeid(*this) != typeid(p))
    throw InvalidOperation();

  Polynom *aux=new Polynom();

  for(int j=0; j<list.size(); ++j)
    {
      Polynom tmp = static_cast<const Polynom&>(p).molt(this->list[j]);
      aux = *aux + tmp;
    }
  aux->setName("Result");
  return aux;
}

Polynom* Polynom::operator/(const BaseClass& p)
{
  Polynom a = static_cast<const Polynom&>(p);

  if(typeid(*this) != typeid(p))
    throw InvalidOperation();

  if(list[0].second < a.list[0].second)
    {
      Polynom *tmp= new Polynom();
      tmp->list.push_back(qMakePair(0,0));
      tmp->setName("Result");
      return tmp;
    }

  int i=0;
  while(i<list.size())
    {
       if(i+1<list.size() && list[i].second-list[i+1].second != 1)
         list.insert(i+1,qMakePair(0,list[i].second-1));
        i++;
    }

  QVector<QPair<double,int>> q;
  while(a.list[0].second<=this->list[0].second)
    {
      q.append(qMakePair(list[0].first/a.list[0].first,list[0].second-a.list[0].second));
       *this = *(*this - a.molt(q.last()));

    }

  if(list.empty())
    list.push_back(qMakePair(0,0));

  this->setName("Resto");


  return new Polynom(q,"Result");
}

Polynom* Polynom::operator=(const BaseClass& p)
{
  if(this != &static_cast<const Polynom&>(p))
    {
      delete this;
      list = static_cast<const Polynom&>(p).list;
    }
  return this;
}

QString Polynom::print() const
{
  QString aux;
  for(int i=0; i<list.size(); ++i)
    {
      if(list[i].first >= 0.0 && i != 0) //aggiunge + se coefficiente positivo
        aux.append("+");
      if(list[i].second == 0) // esponente = 0 non aggungo la x
        aux.append(QString::number(list[i].first));
      else //monomio del tipo ax^b
        {
          if(list[i].first == -1 )
            aux.append('-');
          else
            if(list[i].first != 1.0) //se coefficiente==1 non lo considero
              aux.append(QString::number(list[i].first));

          aux.append("x");
          if(list[i].second != 1) // aggiungo l'esponente se != da 1
            aux.append("^"+QString::number((list[i].second)));
        }
    }
  return aux;
}

